### ISSUE TITLE
Full Unicode support incomplete

### ISSUE THREAD

**Issue by stefanocoretta** [ajaxorg/ace#5431]
### Describe the bug

This is an issue that has been brought up several times through the years but it still not fixed (see for example #1531).

The editor does not handle certain Unicode characters correctly, for example the `Combining Diacritical Marks`. It seems the editor treats these characters as full width characters and the text cursor moves further away from the text.

I described the problem in details here, including pictures: https://github.com/rstudio/rstudio/issues/12680#issue-1575708046 RStudio uses the Ace editor.

### Expected Behavior

Correct handling of Unicode characters.

### Current Behavior

See description.

### Reproduction Steps

See description.

### Ace Version / Browser / OS / Keyboard layout

Ace Kitchen Sink, Safari, macOS; also embedded editor in RStudio latest

**Comment by whazor** [comment]
Hi, I verified that the characters `ã̤n̺ ɾ̺ub̰ĩ̈`, which are monospaced characters get an incorrect double space.

I wonder if this issue would be different from https://github.com/ajaxorg/ace/issues/1153

**Comment by stefanocoretta** [comment]
Yes, it's related to that. :) Do let me know if I can help with testing or the like. I'm happy to provide you with more details about characters/Unicode codes etc if it can help.

**Comment by bradrn** [comment]
Any progress on this issue? It’s severely affecting my application https://bradrn.com/brassica, for which support for combining diacritics is very important.

**Comment by ewnjtgouierg** [comment]
guys, which issue is incorrect cursor placement when there're unicode characters of different width?

this one #5431 or #4142 or #1153 ?

I'm having working fix - (available in my fork for testing) - but i haven't added the fix to the selection yet (selection should use same adjustment)

(if someone is familiar with Selection class enough to point me to where it calculates the coordinates, by the way - you could help me to do it sooner)

**Comment by ewnjtgouierg** [comment]
i fixed this one (combining diacritics) for cursor positioning too, but ace still detects the end of line incorrectly, so if you just press right arrow you get stuck at EOL until you hit one for each diacritic lol

gotta fix that

is there anyone who's been working on ace for a while around? it could really make it all faster if i could consult with someone about what in ace is done where and how, instead of reverse-engineering everything

**Comment by ewnjtgouierg** [comment]
@nightwing , thank you again for being willing to help

as you know, Cursor::getPixelPosition() just multiples column number on a fixed width
which doesn't work for wider characters like emojis

this i fixed (returning actual position instead)

however, for CDM, ace treating diacritics as separate characters - so it calculates the column number itself incorrectly, as well as the line length, so the navigation using arrow keys and END key doesn't work correctly:

- arrow key has to be pressed an extra time for every diacritic

	to fix - replaced .length in Selection::moveCursorRight() with .lengthAce - returns String length ignoring diacritics 

- END key goes to a non-existent column beyond the EOL

	so, I'm checking where does it get the last column:

	it calls Session::getDocumentLastRowColumnPosition()
	which calls Session::screenToDocumentPosition()
	which calls Session::$getStringScreenWidth()

great, Session::$getStringScreenWidth() seems to calculate some kind of a real string length, so it makes sense to adjust it to ignore diacritics

so i do. but it is also called by Session::documentToScreenPosition()
and returning corrected string length there messes up navigation with left/right arrows!

so right now i made Session::$getStringScreenWidth() ignore diacritics only when called by ::screenToDocumentPosition()

it works, but i really-really suspect it... erm... will result in some bugs :D

so, any advises?

my work is here:[
https://github.com/ewnjtgouierg/ace/tree/unicode_pos_fix

**Comment by ewnjtgouierg** [comment]
i've been working on Selection and half way i decided it is just wrong

column must be column, there must not be "wide" characters that are counted as two columns

if there needs to be some columns that depend on character width - then this should be a separate concept, because right now it just unnecessarily overcomplicates everything

so, i've started to moving to 1 charcter = 1  column. but i need to get back to work and stuff so this will be postponed fufufu

meanwhile everyone welcome to comment on this

**Comment by ewnjtgouierg** [comment]
moved to 1 character = 1 column
cursor and selection within one line work for multi-word characters and diacritics

https://github.com/ewnjtgouierg/ace/tree/Selection_Fix

2do: multiline selection and test with line folding and such

will take a pause for a while

**Comment by ewnjtgouierg** [comment]
multi-line selection works, working on word-wrap mode

**Comment by afanasevkirill** [comment]
> multi-line selection works, working on word-wrap mode
> 
> update: word-wrap works
> 
> next: fixing selecting words (using ctrl+shift+arrows) after emojis

Hi! Could I help you somehow to complete this fix?

**Comment by ewnjtgouierg** [comment]
@afanasevkirill

hi! thank you for reaching out

i most likely will get back to this - b/c all my work i do in ace)))
but at the moment i'm a bit overwhelmed by other things :D so not before at least 1 month from now, may be 2

so if you are motivated - feel free to fork and try to finish it - fixing the last thing i outlined, or may be test and find something else you think more important and prefer to fix

i suspect the one utility function i added there became historically too crazy and needs refactoring, sorry in advance if you feel outraged reading that particular code... .))) if you will have questions i'll try to answer

also you'll notice there's unicode.html added there for testing in demo directory... i think that's what i used for testing... probably? )))

all the work is on Selection_Fix branch

p.s. turned on the "Issues" to post there bugs related to this  fix and added you for the case it helps to track the work

**Comment by sKopheK** [comment]
hello, just a small reminder that the issue still exists :)
